### PR TITLE
ARROW-7008: [C++] Check binary offsets and data buffers for nullness in validation. Produce valid arrays in DictionaryEncode on zero-length arrays

### PR DESCRIPTION
**Pull request by wesm**
A couple of related issues:

* Validate() and ValidateFull() did not check the offsets and data buffers of BinaryArray for nullness. We might relax the requirements of the columnar format docs to allow them to be null, but that should be a separate matter
* DictionaryEncode (and other hash kernels) were producing invalid dictionaries for zero-length BinaryArray input

### CODE REVIEW

File: python/pyarrow/tests/test_array.py [hunk 1]
```diff
@@ -1168,6 +1168,15 @@ def test_dictionary_encode_sliced():
         assert result.type == expected.type
 
 
+def test_dictionary_encode_zero_length():
+    # User-facing experience of ARROW-7008
+    arr = pa.array([], type=pa.string())
+    encoded = arr.dictionary_encode()
+    assert len(encoded.dictionary) == 0
+    encoded.validate()
+    encoded.validate(full=True)
```
Comment (pitrou) [inline python/pyarrow/tests/test_array.py:1177]: `validate(full=True)` is a superset of `validate()`, so you don't need both.

File: cpp/src/arrow/compute/kernels/hash_test.cc [hunk 2]
```diff
@@ -644,6 +648,18 @@ TEST_F(TestHashKernel, ChunkedArrayInvoke) {
...
+TEST_F(TestHashKernel, ZeroLengthDictionaryEncode) {
+  // ARROW-7008
+  auto values = ArrayFromJSON(utf8(), "[]");
+  Datum datum_result;
+  ASSERT_OK(DictionaryEncode(&this->ctx_, values, &datum_result));
+
+  std::shared_ptr<Array> result = datum_result.make_array();
+  const auto& dict_result = checked_cast<const DictionaryArray&>(*result);
+  ASSERT_OK(dict_result.Validate());
+  ASSERT_OK(dict_result.ValidateFull());
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/hash_test.cc:660]: `ValidateFull()` is a superset of `Validate()`.

**Review by pitrou — COMMENTED** [review]
LGTM. Just two minor comments.

**Comment by wesm** [comment]
Merging

The Appveyor failure I see https://ci.appveyor.com/project/wesm/arrow/builds/31910981 is reported as https://issues.apache.org/jira/browse/ARROW-8317

**Pull request closed without merging**